[serve][llm] An example for Integrating SGLang with Ray Serve LLM

kangwangamd opened the pull request:
## Description
Integrate SGLang with Ray Serve LLM as an example. 

## Related issues
- Related to #51313

## Additional information
Create a build_sglang_openai_app, similar to the default build_openai_app that uses vLLM. 
Implement OpenAI compatible /v1/chat/completions and /v1/completions endpoints.
usage examples: please go to ray/llm/examples/readme.md

Design doc: https://docs.google.com/document/d/1FLaormOPANCCLBI_UBJLwPl6a9njIyT-kSJfhxLERXA/edit?tab=t.0

kangwangamd committed:
add sglang engine demo for /v1/chat/completions

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
enable engine kwargs to engine via LLMConfig and build_sglang_openai_app

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
enable engine kwargs to engine via LLMConfig and build_sglang_openai_app

Signed-off-by: Colin Wang <kangwang@amd.com>

eicherseiji commented on python/ray/llm/_internal/serve/engines/sglang/sglang_engine.py:
```diff
@@ -0,0 +1,53 @@
+        )
+        return {"echo": res}
+
+    async def llm_config(self) -> Optional[LLMConfig]:
```
self.llm_config is being defined twice here, one function, one object. Try making them unique to fix the TypeError mentioned over Slack

eicherseiji reviewed (commented):
Nice progress! Please fix Cursor/Gemini comments.

kangwangamd committed:
Update sglang_engine.py

remove llm_config as a function

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
workable version of Ray Serve LLM + SGLang PRD

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix /v1/completions and get_deployment_options

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Create sglang-compatibility.md

add sglang-compatibility.md

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix endpoint v1/chat/completions

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
add test_sglang_serve_integration.py

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update builder.py

edit out some comments

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

remove vllm example

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix the SGLang attempts to register system signal handlers issue

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

remove signal issue due to it is resolve by previous commit

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd commented on python/ray/llm/_internal/serve/engines/sglang/sglang_engine.py:
```diff
@@ -0,0 +1,51 @@
+            model_path = "/scratch2/huggingface/hub/meta-llama/Llama-3.1-8B-Instruct/",
+            mem_fraction_static = 0.8,
+            tp_size = 8,
+        )
```
it was fixed

kangwangamd committed:
Update sglang_engine.py

removed redundant import of request

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

change example model to "unsloth/Llama-3.1-8B-Instruct"

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update release_tests.yaml

pytest -sv test_sglang_serve_integration.py

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang_engine.py

fix Potential runtime error on None runtime_env value on get_deploy_option and one line : coding style

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update test_sglang_serve_integration.py

FIX: Removed the incorrect @pytest.mark.asyncio decorator from the synchronous fixture.

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix test model_path for sgl

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

add RAY_EXPERIMENTAL_NOSET_HIP_VISIBLE_DEVICES env-var

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

change default model name

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

fix typo

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

fix missing ','

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

fix typo add MI300X example

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

eicherseiji commented on release/llm_tests/serve/test_sglang_serve_integration.py:
```diff
@@ -0,0 +1,131 @@
+        return False
+
+@pytest.mark.parametrize("model_name", ["hmellor/Ilama-3.2-1B"])
+def test_transcription_model(model_name):
```
Is this a transcription model? Also, please add a test query to this test

eicherseiji commented:
Please sign commits for [DCO](https://github.com/ray-project/ray/pull/58366/checks?check_run_id=56026886096) and check [docs/readthedocs.com:anyscale-ray](https://app.readthedocs.com/projects/anyscale-ray/builds/3541526/) failure. Thanks @kangwangamd!

eicherseiji commented on doc/source/serve/llm/user-guides/sglang-compatibility.md:
```diff
@@ -0,0 +1,96 @@
+    max_tokens=64,
+)
+
+print(f"Response: {response}")
```
<img width="1840" height="1191" alt="Screenshot 2025-11-21 at 11 11 08 AM" src="https://github.com/user-attachments/assets/112dcdd7-193f-43d5-9a1d-7e61c91e2205" />

Example works after changing config to TP1
Example works

kangwangamd committed:
Update release_tests.yaml

change to py 3.11

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update test_sglang_serve_integration.py

add request test to transcription_model

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Update sglang-compatibility.md

remove `APIIngress` (**Frontend**): A FastAPI ingress deployment that listens on HTTP ...

Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>
Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd reviewed (commented):
lgtm

kangwangamd committed:
change to get the value from ray_actor_options

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
remove the release test fro  test_sglang_serve_integration

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
remove release/llm_tests/serve/test_sglang_serve_integration.py

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
create /ray/llm/example/sglang and put sglang-compatibility.md in it

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd commented on python/ray/llm/_internal/serve/engines/sglang/sglang_engine.py:
```diff
@@ -0,0 +1,334 @@
+
+        original_signal_func = signal.signal
+
+        def noop_signal_handler(sig, action):
```
This error occurs because **SGLang** attempts to register system signal handlers (like `SIGQUIT`) to handle graceful shutdowns. However, Python only allows signal handlers to be registered in the **main thread** of the main interpreter.

When running inside **Ray Serve**, your code is running inside a Worker thread (or an Actor that doesn't have main-thread privileges in the way SGLang expects), causing the `ValueError`

**The Fix**:
You need to temporarily "monkey patch" Python's signal.signal function inside your `SGLangServer.__init__` method to do nothing while `sglang.Engine` initializes. This trick fools SGLang into thinking it registered the signal, preventing the crash.

`class SGLangServer:
    def __init__(self, _llm_config: LLMConfig):

        self._llm_config = _llm_config
        self.engine_kwargs = _llm_config.engine_kwargs

        try:
            import sglang
        except ImportError as e:
            raise ImportError(
                "SGLang is not installed. Please run `pip install sglang[all]`."
            ) from e

        # ### FIX START: Monkey patch signal.signal ###
        # SGLang tries to set signal handlers, which fails in Ray Serve workers.
        # We temporarily replace signal.signal with a function that does nothing.
        
        original_signal_func = signal.signal

        def noop_signal_handler(sig, action):
            pass

        try:
            # Override signal.signal with our no-op function
            signal.signal = noop_signal_handler
            
            # Now initialize the engine. SGLang will call signal.signal, 
            # but it will hit our no-op function instead of crashing.
            self.engine = sglang.Engine(**self.engine_kwargs)
            
        finally:
            # STRICTLY restore the original signal function immediately after
            signal.signal = original_signal_func
            
        # ### FIX END ###`

eicherseiji reviewed (commented):
Aligned with @kouroshHakha , please move code to `ray/llm/examples/sglang`. Thanks @kangwangamd!

kangwangamd committed:
remove SGLang from llm_config LLMEngine

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
move sglang-compatibility.md back to doc/source/serve/llm/user-guides

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd commented on doc/source/serve/llm/user-guides/sglang-compatibility.md:
```diff
@@ -0,0 +1,95 @@
+# SGLang Compatibility
```
done

kangwangamd committed:
move sglang related files to ray/llm/examples/sglang, and fix related import path changes

Signed-off-by: Colin Wang <kangwang@amd.com>

nrghosh reviewed (commented):
thanks for addressing comments @kangwangamd 

run pre-commit / lint to unblock test suites from running:
> ./ci/lint/lint.sh pre_commit code_format

kangwangamd committed:
lint remove deployment_cls in args

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
remove log file colin_code_format_check.log

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
remove log file colin_code_format_check.log

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix Handle empty list to prevent IndexError

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix Return SIG_DFL (default handler) to simulate a valid return value for signal.signal()

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix Incorrect request and response types mismatched with OpenAiIngress

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix Gate the setup hook with ENABLE_WORKER_PROCESS_SETUP_HOOK

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix Gate the setup hook with import ENABLE_WORKER_PROCESS_SETUP_HOOK

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd commented on python/ray/llm/examples/sglang/sglang_engine.py:
```diff
@@ -0,0 +1,278 @@
+            self.engine = sglang.Engine(**self.engine_kwargs)
+        finally:
+            signal.signal = original_signal_func
+
```
I think SGLangServer  currently does not need to follow the protocol methods defined in LLMServerProtocol, LLMServerProtocol is for vllm engine. However, we can review what protocols is needed for SGLangServer for better compatible with "v"LLMServerProtocol.

kangwangamd committed:
Merge branch 'master' into feature/serve-sglang

kangwangamd committed:
disable ENABLE_WORKER_PROCESS_SETUP_HOOK for lint pre commit check

Signed-off-by: Colin Wang <kangwang@amd.com>

eicherseiji committed:
Replace docs with literalinclude, add __init__.py

Signed-off-by: Seiji Eicher <seiji@anyscale.com>

eicherseiji committed:
Add warning

Signed-off-by: Seiji Eicher <seiji@anyscale.com>

eicherseiji reviewed (approved):
<img width="2032" height="1167" alt="Screenshot 2025-12-03 at 1 07 06 PM" src="https://github.com/user-attachments/assets/e55571ea-c4aa-4532-901f-5c1ab21c81d6" />

Tested with vllm==0.11.2 and sglang==0.5.6

eicherseiji committed:
Lint

Signed-off-by: Seiji Eicher <seiji@anyscale.com>

kouroshHakha commented on python/ray/llm/examples/sglang/ingress_builder.py:
```diff
@@ -0,0 +1,134 @@
+logger = get_logger(__name__)
+
+
+class IngressClsConfig(BaseModelExtended):
```
why do we need to duplicate this class from https://github.com/ray-project/ray/blob/b0320b19d45c517e36d680f51cbcf727400a144c/python/ray/llm/_internal/serve/core/ingress/builder.py#L25-L43

kouroshHakha commented on python/ray/llm/examples/sglang/ingress_builder.py:
```diff
@@ -0,0 +1,134 @@
+        return self
+
+
+def build_sglang_openai_app(builder_config: dict) -> Application:
```
I feel like we just need this custom builder as the main entrypoint?

kouroshHakha commented on python/ray/llm/examples/__init__.py:
Examples should not be part of the package. We should just create a simple yaml that locally uses the modules under the examples/sglang folder. 

So sth like this:

```
examples/sglang/modules/ingress_builder.py etc.
examples/sglang/serve_xyz.yaml 
examples/sglang/readme.md
```

kouroshHakha reviewed (commented):
Hey @kangwangamd thanks for your contribution and sorry for my late reply. I was OOO. left some comments that are mostly around code restructuring and import formats to better position this as an example rather than part of ray package. Thanks.

kangwangamd commented on python/ray/llm/examples/sglang/ingress_builder.py:
```diff
@@ -0,0 +1,134 @@
+        return value
+
+
+class LLMServingArgs(BaseModelExtended):
```
you are right

kangwangamd committed:
modularized the core generation and response parsing into the private method _generate_and_extract_metadata

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
move and merge ingress_builder and server_builder

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
add a serve_sglang_example.py

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
remove files ingress_builder and server_builder

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix Calling strip() on None and Empty prompt list causes IndexError crash error

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
move sglang-compatibility.md to llm/examples/readme.md

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
change import SGLangServer to relative path

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
lint fix

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
lint fix

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix completions batch issue

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix lint

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
lint

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Merge branch 'master' into feature/serve-sglang

kangwangamd committed:
Merge branch 'master' into feature/serve-sglang

eicherseiji commented on python/ray/llm/examples/serve_sglang_example.py:
```diff
@@ -0,0 +1,26 @@
+from ray import serve
+from ray.llm.examples.sglang.modules.builder import build_sglang_openai_app
```
```suggestion
from modules.builder import build_sglang_openai_app
```

kouroshHakha replied:
+1

kangwangamd committed:
Update python/ray/llm/examples/sglang/modules/builder.py

thx for fix this

Co-authored-by: Seiji Eicher <58963096+eicherseiji@users.noreply.github.com>
Signed-off-by: kangwangamd <100359556+kangwangamd@users.noreply.github.com>

kangwangamd committed:
remove init.py and move serve_sglang_example.py inside examples/sglang

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
lint

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
move sglang_engine file path

Signed-off-by: Colin Wang <conlinwang@ntu.edu.tw>

kangwangamd committed:
Merge branch 'master' into feature/serve-sglang

kouroshHakha commented on python/ray/llm/examples/sglang/modules/builder.py:
```diff
@@ -0,0 +1,111 @@
+    )
+
+
+def build_sglang_openai_app(builder_config: dict) -> Application:
```
> Instead of creating a separate build_sglang_deployment we should use build_llm_deployment directly. In that utiltiy you can pass in the target class. So this method definition here feels redundant.

same comment as above about building the ingress. The existing build_openai_app can be and actually should be reused, because that utility can be modified so that you can plug and play your own LLMServer deployment? 


Basically in LLMConfig we can add this attribute:

````
class LLMConfig:

       llm_server_cls: Union[str, Type] = None

````

and in the build method we can just read LLMConfig for this attribute and use that if provided. This will make LLMConfig pluggable from LLMServer class perspective. You can then remove all these builder functionality because you can reuse the same build_openai_app then? 

If I am not missing anything we should move towards that direction.

kouroshHakha commented on python/ray/llm/examples/sglang/serve_sglang_example.py:
```diff
@@ -0,0 +1,26 @@
+from ray import serve
+from ray.llm.examples.sglang.modules.builder import build_sglang_openai_app
```
I don't think this import path works anymore since we removed the init files. It should be relative now. 

```
from modules.builder import build_sglang_openai_app
```

kangwangamd committed:
fix lint

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
remove sgl_pi_run.log accidental local execution artifact

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
remove print log

Signed-off-by: Colin Wang <conlinwang@ntu.edu.tw>

kangwangamd committed:
remove sglang/modules/builder use core/ingree/builder

Signed-off-by: Colin Wang <kangwang@amd.com>

kouroshHakha reviewed (commented):
Just one renaming thing that is subjective but I think makes sense at this time. I think it looks good o.w.

kangwangamd committed:
rename deployment_cls to server_cls

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
edit readme for what is not support in this example for milestone 1

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
fix rename deployment_cls

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Merge branch 'ray-project:master' into feature/serve-sglang

kangwangamd committed:
move sglang readme file into example sglang folder

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Merge branch 'master' into feature/serve-sglang

eicherseiji reviewed (commented):
Please remove the extra GPU resource here: https://github.com/ray-project/ray/pull/58366/files#r2630345320

And then we're good to merge from my side as well.

kangwangamd committed:
fix default placement group requests 2 GPUs for single-GPU example

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
XMerge branch 'feature/serve-sglang' of https://github.com/kangwangamd/ray into feature/serve-sglang

kangwangamd committed:
add TO DO expression on tp_size > 1 cases

Signed-off-by: Colin Wang <kangwang@amd.com>

kangwangamd committed:
Merge branch 'master' into feature/serve-sglang

kangwangamd committed:
Merge branch 'master' into feature/serve-sglang

eicherseiji reviewed (approved):
The engine serves a request with `RAY_EXPERIMENTAL_NOSET_CUDA_VISIBLE_DEVICES=0`, `sglang==0.5.6.post2`, on top of `anyscale/ray-llm:nightly-py311-cu128`.

<img width="1840" height="1191" alt="Screenshot 2025-12-19 at 11 02 19 AM" src="https://github.com/user-attachments/assets/31443512-da22-41e3-ae90-b8a99d84803c" />

Ready to merge

kouroshHakha reviewed (approved):
Nice work @kangwangamd 🚀
